File: sdks/ts/packages/golem-ts-sdk/src/http/endpoint.ts
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Golem Source License v1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TypeMetadata } from '@golemcloud/golem-ts-types-core';
import {
  AuthDetails,
  CorsOptions,
  HeaderVariable,
  HttpEndpointDetails,
  PathSegment,
  QueryVariable,
} from 'golem:agent/common';
import { parsePath } from './path';
import { parseQuery } from './query';
import { AgentMethodRegistry } from '../internal/registry/agentMethodRegistry';

export type EndpointDecoratorOptions = {
  get?: string;
  post?: string;
  put?: string;
  delete?: string;
  custom?: string;
  headers?: Record<string, string>;
  auth?: boolean;
  cors?: string[];
};

/**
 * Decorates a method as an HTTP endpoint for this agent.
 */
export function endpoint(opts: EndpointDecoratorOptions) {
  return function (
    target: Object,
    propertyKey: string | symbol,
    descriptor: PropertyDescriptor,
  ) {
    const className = target.constructor.name;
    const methodName = String(propertyKey);

    const classMetadata = TypeMetadata.get(className);
    if (!classMetadata) {
      throw new Error(
        `Class metadata not found for agent ${className}. Ensure metadata is generated.`,
      );
    }

    const pathAndQuery = split_path_and_query(
      opts.get || opts.post || opts.put || opts.delete || opts.custom || '',
    );

    const path = pathAndQuery.path;
    const query = pathAndQuery.query;

    let httpMethod: HttpEndpointDetails['httpMethod'];

    if (opts.get) {
      httpMethod = { tag: 'get' };
    } else if (opts.post) {
      httpMethod = { tag: 'post' };
    } else if (opts.put) {
      httpMethod = { tag: 'put' };
    } else if (opts.delete) {
      httpMethod = { tag: 'delete' };
    } else if (opts.custom) {
      httpMethod = { tag: 'custom', val: opts.custom };
    } else {
      throw new Error(
        `Endpoint decorator must specify one of get/post/put/delete/custom for method ${methodName}`,
      );
    }

    if (!path.startsWith('/')) {
      throw new Error(
        `Endpoint path must start with '/'. Method: ${methodName}`,
      );
    }

    const pathSuffix: PathSegment[] = parsePath(path);

    const headerVars: HeaderVariable[] = Object.entries(opts.headers || {}).map(
      ([headerName, variableName]) => ({ headerName, variableName }),
    );

    const queryVars: QueryVariable[] = query ? parseQuery(query) || [] : [];
    const authDetails: AuthDetails = { required: opts.auth ?? true };
    const corsOptions: CorsOptions = { allowedPatterns: opts.cors ?? [] };

    const httpEndpoint: HttpEndpointDetails = {
      httpMethod,
      pathSuffix,
      headerVars,
      queryVars,
      authDetails,
      corsOptions,
    };

    AgentMethodRegistry.setHttpEndpoint(className, methodName, httpEndpoint);
  };
}

function split_path_and_query(pathWithQuery: string): {
  path: string;
  query: string | null;
} {
  const [path, query] = pathWithQuery.split('?', 2);
  return { path, query: query ?? null };
}

File: sdks/rust/golem-rust/src/agentic/http/endpoint.rs
// Copyright 2024-2025 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::agentic::http::path::parse_path;
use crate::agentic::http::query::parse_query;
use crate::golem_agentic::golem::agent::common::{
    AuthDetails, CorsOptions, HeaderVariable, HttpEndpointDetails, HttpMethod,
};

pub fn get_http_endpoint_details(
    method: &str,
    path: &str,
    auth: Option<bool>,
    cors_options: Vec<String>,
    http_headers: Vec<(String, String)>,
) -> Result<HttpEndpointDetails, String> {
    let PathAndQuery { path, query } = split_path_and_query(path);

    let path_suffix = parse_path(&path).map_err(|e| e.to_string())?;

    let query_vars = match query {
        Some(q) => parse_query(q.as_str()).map_err(|e| e.to_string())?,
        None => vec![],
    };

    let http_method = match method {
        "get" => HttpMethod::Get,
        "post" => HttpMethod::Post,
        "put" => HttpMethod::Put,
        "delete" => HttpMethod::Delete,
        "patch" => HttpMethod::Patch,
        "head" => HttpMethod::Head,
        "options" => HttpMethod::Options,
        "connect" => HttpMethod::Connect,
        "trace" => HttpMethod::Trace,
        other => return Err(format!("Unsupported HTTP method: {}", other)),
    };

    let header_vars: Vec<HeaderVariable> = http_headers
        .into_iter()
        .map(|(header_name, variable_name)| HeaderVariable {
            header_name,
            variable_name,
        })
        .collect();

    Ok(HttpEndpointDetails {
        http_method,
        path_suffix,
        header_vars,
        query_vars,
        auth_details: auth.map(|required| AuthDetails { required }),
        cors_options: CorsOptions {
            allowed_patterns: cors_options,
        },
    })
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct PathAndQuery {
    path: String,
    query: Option<String>,
}

fn split_path_and_query(path_with_query: &str) -> PathAndQuery {
    let mut parts = path_with_query.splitn(2, '?');

    PathAndQuery {
        path: parts.next().unwrap_or("").to_string(),
        query: parts.next().map(|q| q.to_string()),
    }
}
